File: arrays/03_restore_from_pairs/restore_from_pairs.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cmath>
#include <stdexcept>

using namespace std;


class Solution {
public:
    vector<int> restoreFromAdjacentPairs(const vector<vector<int>>& adjacentPairs) {
        if (!checkConstraints(adjacentPairs)) {
            throw std::runtime_error("Constraints violated");
        }

        unordered_map<int, vector<int>> adjacencyList;
        for (const auto& p : adjacentPairs) {
            adjacencyList[p[0]].push_back(p[1]);
            adjacencyList[p[1]].push_back(p[0]);
        }

        vector<int> nums;
        for (const auto& p : adjacencyList) {
            if (p.second.size() == 1) {
                nums.push_back(p.first);
                break;
            }
        }
        nums.push_back(adjacencyList[nums[0]][0]);

        while (nums.size() < adjacencyList.size()) {
            int last = nums.back();
            int next = (adjacencyList[last][0] != nums[nums.size() - 2]) ? adjacencyList[last][0] : adjacencyList[last][1];
            nums.push_back(next);
        }

        return nums;
    }

private:
    // Helper function to check the constraints
    bool checkConstraints(const vector<vector<int>>& adjacentPairs) {
        int n = adjacentPairs.size() + 1; 

        if (n < 2 || n > pow(10, 5)) return false;

        for (int i = 0; i < n - 1; ++i) {
            if (adjacentPairs[i].size() != 2) return false;
            if (adjacentPairs[i][0] < pow(-10, 5) || adjacentPairs[i][0] > pow(10, 5)) return false;
            if (adjacentPairs[i][1] < pow(-10, 5) || adjacentPairs[i][1] > pow(10, 5)) return false;
        }

        return true;
    }
};

File: arrays/03_restore_from_pairs/restore_from_pairs.py
from typing import List
from collections import defaultdict


class Solution:
    def restore_from_adjacent_pairs(self, adjacent_pairs: List[List[int]]) -> List[int]:
        if not self._check_constraints(adjacent_pairs):
            raise ValueError("Constraints violated")

        adjacency_list = defaultdict(list)
        for u, v in adjacent_pairs:
            adjacency_list[u].append(v)
            adjacency_list[v].append(u)

        nums: List[int] = list()
        for key, value in adjacency_list.items():
            if len(value) == 1:
                nums.append(key)
                nums.append(value[0])
                break

        while len(nums) < len(adjacency_list):
            last = nums[-1]
            next = adjacency_list[last][0] if adjacency_list[last][0] != nums[-2] else adjacency_list[last][1]
            nums.append(next)

        return nums

    def _check_constraints(self, adjacent_pairs: List[List[int]]) -> bool:
        n = len(adjacent_pairs) + 1
        if n < 2 or n > 10**5:
            return False

        for i in range(n):
            if len(adjacent_pairs[i]) != 2:
                return False
            if adjacent_pairs[i][0] < -10**5 or adjacent_pairs[i][0] > 10**5:
                return False
            if adjacent_pairs[i][1] < -10**5 or adjacent_pairs[i][1] > 10**5:
                return False

        return True
